rls/src/main/java/io/grpc/rls/internal/LinkedHashLruCache.java

```diff
@@ -0,0 +1,381 @@
+    return true;
+  }
+
+  /** Determines if the entry is already expired or not. */
+  protected abstract boolean isExpired(K key, V value, long nowNanos);
+
+  /**
+   * Returns estimated size of entry to keep track. If it always returns 1, the max size bytes
+   * behaves like max number of entry (default behavior).
+   */
+  @SuppressWarnings("unused")
+  protected int estimateSizeOf(K key, V value) {
+    return 1;
+  }
+
+  /** Updates size for given key if entry exists. It is useful if the cache value is mutated. */
+  public void updateEntrySize(K key) {
+    SizedValue entry = readInternal(key);
+    if (entry == null) {
+      return;
+    }
+    int prevSize = entry.size;
+    int newSize = estimateSizeOf(key, entry.value);
+    entry.size = newSize;
+    estimatedSizeBytes.addAndGet(newSize - prevSize);
+  }
+
+  @Override
+  @Nullable
+  public final V cache(K key, V value) {
+    checkNotNull(key, "key");
+    checkNotNull(value, "value");
+    SizedValue existing;
+    int size = estimateSizeOf(key, value);
+    synchronized (lock) {
+      estimatedSizeBytes.addAndGet(size);
+      existing = delegate.put(key, new SizedValue(size, value));
+      if (existing != null) {
+        evictionListener.onEviction(key, existing, EvictionType.REPLACED);
+      }
+    }
+    return existing == null ? null : existing.value;
+  }
+
+  @Override
+  @Nullable
+  @CheckReturnValue
+  public final V read(K key) {
+    SizedValue entry = readInternal(key);
+    if (entry != null) {
+      return entry.value;
+    }
+    return null;
+  }
+
+  @Nullable
+  @CheckReturnValue
+  private SizedValue readInternal(K key) {
+    checkNotNull(key, "key");
+    synchronized (lock) {
+      SizedValue existing = delegate.get(key);
+      if (existing != null && isExpired(key, existing.value, timeProvider.currentTimeNanos())) {
+        invalidate(key, EvictionType.EXPIRED);
+        return null;
+      }
+      return existing;
+    }
+  }
+
+  @Override
+  @Nullable
+  public final V invalidate(K key) {
+    return invalidate(key, EvictionType.EXPLICIT);
+  }
+
+  @Nullable
+  private V invalidate(K key, EvictionType cause) {
+    checkNotNull(key, "key");
+    checkNotNull(cause, "cause");
+    synchronized (lock) {
+      SizedValue existing = delegate.remove(key);
+      if (existing != null) {
+        evictionListener.onEviction(key, existing, cause);
+      }
+      return existing == null ? null : existing.value;
+    }
+  }
+
+  @Override
+  public final void invalidateAll(Iterable<K> keys) {
+    checkNotNull(keys, "keys");
+    synchronized (lock) {
+      for (K key : keys) {
+        SizedValue existing = delegate.remove(key);
+        if (existing != null) {
+          evictionListener.onEviction(key, existing, EvictionType.EXPIRED);
+        }
+      }
+    }
+  }
+
+  @Override
+  @CheckReturnValue
+  public final boolean hasCacheEntry(K key) {
+    // call get to handle expired
```

Comment:
I'm not clear what the comment means. Is it necessary?

Reply:
i can improve the comment
if you get why it call readInternal instead of LinkedHashMap#contains without much thinking i am more than happy to remove it.